File: gitlfs-client/src/main/java/ru/bozaro/gitlfs/client/internal/Request.java
package ru.bozaro.gitlfs.client.internal;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.http.HttpResponse;
import org.apache.http.HttpStatus;

import javax.annotation.Nonnull;
import java.io.IOException;

/**
 * Single HTTP request.
 *
 * @author Artem V. Navrotskiy
 */
public interface Request<R> {

  @Nonnull
  LfsRequest createRequest(@Nonnull ObjectMapper mapper, @Nonnull String url) throws IOException;

  R processResponse(@Nonnull ObjectMapper mapper, @Nonnull HttpResponse response) throws IOException;

  /**
   * Success status codes.
   *
   * @return Success status codes.
   */
  @Nonnull
  default int[] statusCodes() {
    return new int[]{HttpStatus.SC_OK};
  }
}

File: gitlfs-client/src/main/kotlin/ru/bozaro/gitlfs/client/internal/Request.kt
package ru.bozaro.gitlfs.client.internal

import com.fasterxml.jackson.databind.ObjectMapper
import org.apache.http.HttpResponse
import org.apache.http.HttpStatus
import java.io.IOException

/**
 * Single HTTP request.
 *
 * @author Artem V. Navrotskiy
 */
interface Request<R> {
    @Throws(IOException::class)
    fun createRequest(mapper: ObjectMapper, url: String): LfsRequest

    @Throws(IOException::class)
    fun processResponse(mapper: ObjectMapper, response: HttpResponse): R

    /**
     * Success status codes.
     *
     * @return Success status codes.
     */
    fun statusCodes(): IntArray {
        return intArrayOf(HttpStatus.SC_OK)
    }
}
